src/main/java/org/cyclonedx/gradle/DependencyGraphTraverser.java

```diff
@@ -0,0 +1,221 @@
+    private final MavenProjectLookup mavenLookup;
+    private final boolean includeMetaData;
+    private final MavenHelper mavenHelper;
+
+    DependencyGraphTraverser(
+            final Logger logger,
+            final Map<ComponentIdentifier, File> resolvedArtifacts,
+            final MavenProjectLookup mavenLookup,
+            final CycloneDxTask task) {
+        this.logger = logger;
+        this.resolvedArtifacts = resolvedArtifacts;
+        this.mavenLookup = mavenLookup;
+        this.includeMetaData = task.getIncludeMetadataResolution().get();
+        this.mavenHelper = new MavenHelper(logger, task.getIncludeLicenseText().get());
+    }
+
+    /**
+     * Traverses the dependency graph of a configuration belonging to the specified project
+     *
+     * @param rootNode entry point into the graph which is typically represents a project
+     * @param projectName project to which the configuration belongs to
+     * @param configName name of the configuration
+     *
+     * @return a graph represented as map which is fully serializable. The graph nodes are instances of
+     * SbomComponent which contain the necessary information to generate the Bom
+     */
+    Map<SbomComponentId, SbomComponent> traverseGraph(
+            final ResolvedComponentResult rootNode, final String projectName, final String configName) {
+
+        final Map<GraphNode, Set<GraphNode>> graph = new HashMap<>();
+        final Queue<GraphNode> queue = new ArrayDeque<>();
+
+        final GraphNode rootGraphNode = new GraphNode(rootNode);
+        rootGraphNode.inScopeConfiguration(projectName, configName);
+        queue.add(rootGraphNode);
+
+        logger.debug("CycloneDX: Traversal of graph for configuration {} of project {}", configName, projectName);
+        while (!queue.isEmpty()) {
+            final GraphNode graphNode = queue.poll();
+            if (!graph.containsKey(graphNode)) {
+                graph.put(graphNode, new HashSet<>());
+                logger.debug("CycloneDX: Traversing node with ID {}", graphNode.id);
+                for (final DependencyResult dep : graphNode.getResult().getDependencies()) {
+                    if (dep instanceof ResolvedDependencyResult) {
+                        final ResolvedComponentResult dependencyComponent =
+                                ((ResolvedDependencyResult) dep).getSelected();
+                        logger.debug(
+                                "CycloneDX: Node with ID {} has dependency with ID {}",
+                                graphNode.id,
+                                dependencyComponent);
+                        final GraphNode dependencyNode = new GraphNode(dependencyComponent);
+                        dependencyNode.inScopeConfiguration(projectName, configName);
+                        graph.get(graphNode).add(dependencyNode);
+                        queue.add(dependencyNode);
+                    }
```

Comment:
Should we log any `UnresolvedDependencyResult`?

```suggestion
                    } else if (dep instanceof UnresolvedDependencyResult) {
                        UnresolvedDependencyResult unresolved = (UnresolvedDependencyResult) dep;
                        logger.debug(
                                "Unable to resolve artifact `{}` because {}",
                                unresolved.getAttempted().getDisplayName(),
                                unresolved.getFailure().toString());
                    }
```